core/deployment/src/main/java/io/quarkus/deployment/steps/MainClassBuildStep.java

```diff
@@ -308,6 +333,78 @@ void build(List<StaticBytecodeRecorderBuildItem> staticInitTasks,
         file.close();
     }
 
+    /**
+     * Step timing is enabled only if 'quarkus.debug.print-startup-times' is set.
+     * This method should be called both in static init and runtime init in order to
+     * make sure that the system properly works properly in native mode
+     */
+    private void setStepTimingField(MethodCreator mv) {
+        ResultHandle timingEnabledProp = mv.invokeStaticMethod(
+                ofMethod(System.class, "getProperty", String.class, String.class, String.class),
+                mv.load(MainClassBuildStep.PRINT_STARTUP_TIMES_PROPERTY), mv.load("false"));
+        mv.writeStaticField(STEP_TIMING_ENABLED_FIELD, mv.invokeStaticMethod(
+                ofMethod(Boolean.class, "parseBoolean", boolean.class, String.class), timingEnabledProp));
+    }
+
+    /**
+     * Generates the following bytecode:
+     * 
+     * <pre>
+     *
+     * private static void printStepTime() {
+     *     if (stepTimingEnabled) {
+     *         long var2 = System.currentTimeMillis();
+     *         String var0 = STARTUP_CONTEXT.getCurrentBuildStepName();
+     *         StringBuilder var1 = new StringBuilder("Build step ");
+     *         var1.append(var0);
+     *         var1.append(" completed in: ");
+     *         long var4 = stepTimingStart;
+     *         String var6 = String.valueOf(Math.subtractExact(var2, var4));
+     *         var1.append(var6);
+     *         var1.append("ms");
+     *         PrintStream var7 = System.out;
+     *         String var8 = var1.toString();
+     *         var7.println(var8);
+     *         stepTimingStart = System.currentTimeMillis();
+     *     }
+     * }
+     *
+     * </pre>
+     *
+     * This is used to print the time it took for a build step to execute. The first invocation of this method
+     * assumes that the static {@code stepTimingStart} field has been set.
+     *
+     * After printing the time, it also resets {@code stepTimingStart} because steps are executed one after the other.
+     */
+    private void generationPrintStepTimeMethod(ClassCreator file) {
```

Comment:
Does this actually need to be generated? Can't it just be a static utility method?

Reply:
In it's current form yes because it changes uses the classes static fields. Do you prefer this not be generated and instead have the  `stepTimingEnabled` check and setting of `stepTimingStart` outside the method?

Reply:
Or we could move the fields to the utility class as well :)

Reply:
Yea, I was thinking just move as much as possible to the utility class.

Reply:
On it

Reply:
PR updated